kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/CreateCheckpointIterator.java

```diff
@@ -0,0 +1,366 @@
+
+    private final TableClient tableClient;
+    private final LogSegment logSegment;
+    private final long minFileRetentionTimestampMillis;
+
+    // State of the iterator and current batch being worked on
+    private CloseableIterator<ActionWrapper> actionsIter;
+    private boolean closed;
+    private Optional<FilteredColumnarBatch> toReturnNext = Optional.empty();
+    /**
+     * This buffer is reused across batches to keep the memory allocations minimal. It is resized as
+     * required and the array entries are reset between batches.
+     */
+    private boolean[] selectionVectorBuffer;
+
+    // Current state of the tombstones and add files from delta files
+    private final Set<UniqueFileActionTuple> tombstonesFromJson = new HashSet<>();
+    private final Set<UniqueFileActionTuple> addFilesFromJson = new HashSet<>();
+
+    // Current state of the protocol and metadata. Captures whether protocol or metadata is seen.
+    // We traverse the log in reverse, so the first encounter of protocol or metadata is considered
+    // latest.
+    private boolean isMetadataAlreadySeen;
+    private boolean isProtocolAlreadySeen;
+
+    // Current state of the transaction identifier (a.k.a. SetTransaction). We traverse the log in
+    // reverse, so storing the first seen transaction version for each appId is enough for
+    // checkpoint
+    private final Map<String, Long> txnAppIdToVersion = new HashMap<>();
+
+    // Metadata about the checkpoint to store in `_last_checkpoint` file
+    private long numberOfAddActions = 0; // final number of add actions survived in the checkpoint
+
+    /////////////////
+    // Public APIs //
+    /////////////////
+
+    public CreateCheckpointIterator(
+            TableClient tableClient,
+            LogSegment logSegment,
+            long minFileRetentionTimestampMillis) {
+        this.tableClient = tableClient;
+        this.logSegment = logSegment;
+        this.minFileRetentionTimestampMillis = minFileRetentionTimestampMillis;
+    }
+
+    @Override
+    public boolean hasNext() {
+        initActionIterIfRequired();
+        checkState(!closed, "Can't call `hasNext` on a closed iterator.");
+        return prepareNext();
+    }
+
+    @Override
+    public FilteredColumnarBatch next() {
+        checkState(!closed, "Can't call `next` on a closed iterator.");
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+
+        FilteredColumnarBatch toReturn = toReturnNext.get();
+        toReturnNext = Optional.empty();
+        return toReturn;
+    }
+
+    @Override
+    public void close() throws IOException {
+        closed = true;
+        Utils.closeCloseables(actionsIter);
+    }
+
+    /**
+     * Number of add files in the final checkpoint. Should be called once the entire data of this
+     * iterator is consumed.
+     *
+     * @return Number of add files in checkpoint.
+     */
+    public long getNumberOfAddActions() {
```

Comment:
should we throw or prevent users from calling this before the entire state is consumed?

Reply:
added